# Change Log

## [Unreleased](https://github.com/sanger-pathogens/Fastaq/tree/HEAD)

[Full Changelog](https://github.com/sanger-pathogens/Fastaq/compare/v3.17.0...HEAD)

**Merged pull requests:**

- Update readme [\#73](https://github.com/sanger-pathogens/Fastaq/pull/73) ([ssjunnebo](https://github.com/ssjunnebo))

## [v3.17.0](https://github.com/sanger-pathogens/Fastaq/tree/v3.17.0) (2018-02-19)
[Full Changelog](https://github.com/sanger-pathogens/Fastaq/compare/v3.16.0...v3.17.0)

**Merged pull requests:**

- Added a function to calculate GC content of a sequence [\#68](https://github.com/sanger-pathogens/Fastaq/pull/68) ([mbhall88](https://github.com/mbhall88))

## [v3.16.0](https://github.com/sanger-pathogens/Fastaq/tree/v3.16.0) (2017-11-29)
[Full Changelog](https://github.com/sanger-pathogens/Fastaq/compare/v3.15.0...v3.16.0)

**Merged pull requests:**

- ignore everything after a space in a FASTQ readname line [\#67](https://github.com/sanger-pathogens/Fastaq/pull/67) ([andrewjpage](https://github.com/andrewjpage))
- update LICENSE and AUTHORS [\#66](https://github.com/sanger-pathogens/Fastaq/pull/66) ([ssjunnebo](https://github.com/ssjunnebo))

## [v3.15.0](https://github.com/sanger-pathogens/Fastaq/tree/v3.15.0) (2017-02-10)
[Full Changelog](https://github.com/sanger-pathogens/Fastaq/compare/v3.14.0...v3.15.0)

**Merged pull requests:**

- Sort by name [\#65](https://github.com/sanger-pathogens/Fastaq/pull/65) ([vaofford](https://github.com/vaofford))
- spelling [\#64](https://github.com/sanger-pathogens/Fastaq/pull/64) ([satta](https://github.com/satta))

## [v3.14.0](https://github.com/sanger-pathogens/Fastaq/tree/v3.14.0) (2016-09-01)
[Full Changelog](https://github.com/sanger-pathogens/Fastaq/compare/v3.13.0...v3.14.0)

**Merged pull requests:**

- Acgtn only [\#63](https://github.com/sanger-pathogens/Fastaq/pull/63) ([martinghunt](https://github.com/martinghunt))
- fix spelling [\#62](https://github.com/sanger-pathogens/Fastaq/pull/62) ([satta](https://github.com/satta))

## [v3.13.0](https://github.com/sanger-pathogens/Fastaq/tree/v3.13.0) (2016-08-12)
[Full Changelog](https://github.com/sanger-pathogens/Fastaq/compare/v3.12.1...v3.13.0)

**Merged pull requests:**

- Random reads use float [\#61](https://github.com/sanger-pathogens/Fastaq/pull/61) ([martinghunt](https://github.com/martinghunt))
- Add build status [\#60](https://github.com/sanger-pathogens/Fastaq/pull/60) ([martinghunt](https://github.com/martinghunt))
- Add travis CI [\#59](https://github.com/sanger-pathogens/Fastaq/pull/59) ([martinghunt](https://github.com/martinghunt))

## [v3.12.1](https://github.com/sanger-pathogens/Fastaq/tree/v3.12.1) (2016-04-18)
[Full Changelog](https://github.com/sanger-pathogens/Fastaq/compare/v3.12.0...v3.12.1)

**Merged pull requests:**

- Make  \_\_version\_\_  work [\#58](https://github.com/sanger-pathogens/Fastaq/pull/58) ([martinghunt](https://github.com/martinghunt))

## [v3.12.0](https://github.com/sanger-pathogens/Fastaq/tree/v3.12.0) (2016-04-14)
[Full Changelog](https://github.com/sanger-pathogens/Fastaq/compare/v3.11.1...v3.12.0)

**Merged pull requests:**

- Make gene from seq [\#57](https://github.com/sanger-pathogens/Fastaq/pull/57) ([martinghunt](https://github.com/martinghunt))

## [v3.11.1](https://github.com/sanger-pathogens/Fastaq/tree/v3.11.1) (2016-01-04)
[Full Changelog](https://github.com/sanger-pathogens/Fastaq/compare/v3.11.0...v3.11.1)

**Merged pull requests:**

- Fix looks like gene [\#56](https://github.com/sanger-pathogens/Fastaq/pull/56) ([martinghunt](https://github.com/martinghunt))
- fix spelling occurence -\> occurrence [\#55](https://github.com/sanger-pathogens/Fastaq/pull/55) ([satta](https://github.com/satta))

## [v3.11.0](https://github.com/sanger-pathogens/Fastaq/tree/v3.11.0) (2015-11-30)
[Full Changelog](https://github.com/sanger-pathogens/Fastaq/compare/v3.10.1...v3.11.0)

**Merged pull requests:**

- Interleave add suffix [\#54](https://github.com/sanger-pathogens/Fastaq/pull/54) ([martinghunt](https://github.com/martinghunt))

## [v3.10.1](https://github.com/sanger-pathogens/Fastaq/tree/v3.10.1) (2015-11-16)
[Full Changelog](https://github.com/sanger-pathogens/Fastaq/compare/v3.10.0...v3.10.1)

**Merged pull requests:**

- Qual char option to tiling BAM. Properly remove long\_read\_simulate [\#53](https://github.com/sanger-pathogens/Fastaq/pull/53) ([martinghunt](https://github.com/martinghunt))

## [v3.10.0](https://github.com/sanger-pathogens/Fastaq/tree/v3.10.0) (2015-10-16)
[Full Changelog](https://github.com/sanger-pathogens/Fastaq/compare/v3.9.0...v3.10.0)

**Merged pull requests:**

- Remove numpy [\#51](https://github.com/sanger-pathogens/Fastaq/pull/51) ([martinghunt](https://github.com/martinghunt))

## [v3.9.0](https://github.com/sanger-pathogens/Fastaq/tree/v3.9.0) (2015-10-13)
[Full Changelog](https://github.com/sanger-pathogens/Fastaq/compare/v3.8.0...v3.9.0)

**Merged pull requests:**

- Length offsets from fai [\#50](https://github.com/sanger-pathogens/Fastaq/pull/50) ([martinghunt](https://github.com/martinghunt))

## [v3.8.0](https://github.com/sanger-pathogens/Fastaq/tree/v3.8.0) (2015-09-25)
[Full Changelog](https://github.com/sanger-pathogens/Fastaq/compare/v3.7.0...v3.8.0)

**Merged pull requests:**

- Stats from fai [\#49](https://github.com/sanger-pathogens/Fastaq/pull/49) ([martinghunt](https://github.com/martinghunt))

## [v3.7.0](https://github.com/sanger-pathogens/Fastaq/tree/v3.7.0) (2015-09-01)
[Full Changelog](https://github.com/sanger-pathogens/Fastaq/compare/v3.6.1...v3.7.0)

**Merged pull requests:**

- Boulderio format [\#48](https://github.com/sanger-pathogens/Fastaq/pull/48) ([martinghunt](https://github.com/martinghunt))

## [v3.6.1](https://github.com/sanger-pathogens/Fastaq/tree/v3.6.1) (2015-08-18)
[Full Changelog](https://github.com/sanger-pathogens/Fastaq/compare/v3.6.0...v3.6.1)

**Merged pull requests:**

- setup.py tweaks for numpy and nose [\#47](https://github.com/sanger-pathogens/Fastaq/pull/47) ([martinghunt](https://github.com/martinghunt))

## [v3.6.0](https://github.com/sanger-pathogens/Fastaq/tree/v3.6.0) (2015-07-28)
[Full Changelog](https://github.com/sanger-pathogens/Fastaq/compare/v.3.5.0...v3.6.0)

**Merged pull requests:**

- Check names unique [\#46](https://github.com/sanger-pathogens/Fastaq/pull/46) ([martinghunt](https://github.com/martinghunt))

## [v.3.5.0](https://github.com/sanger-pathogens/Fastaq/tree/v.3.5.0) (2015-04-23)
[Full Changelog](https://github.com/sanger-pathogens/Fastaq/compare/v3.4.0...v.3.5.0)

**Merged pull requests:**

- Intervals add distance\_to\_point function [\#45](https://github.com/sanger-pathogens/Fastaq/pull/45) ([martinghunt](https://github.com/martinghunt))

## [v3.4.0](https://github.com/sanger-pathogens/Fastaq/tree/v3.4.0) (2015-04-20)
[Full Changelog](https://github.com/sanger-pathogens/Fastaq/compare/v3.3.1...v3.4.0)

**Merged pull requests:**

- To fastg [\#44](https://github.com/sanger-pathogens/Fastaq/pull/44) ([martinghunt](https://github.com/martinghunt))

## [v3.3.1](https://github.com/sanger-pathogens/Fastaq/tree/v3.3.1) (2015-04-08)
[Full Changelog](https://github.com/sanger-pathogens/Fastaq/compare/v3.3.0...v3.3.1)

**Merged pull requests:**

- updated version number [\#43](https://github.com/sanger-pathogens/Fastaq/pull/43) ([nds](https://github.com/nds))
- Minor fix in to\_orfs\_gff.py: parser.gff\_out to parser.outfile [\#42](https://github.com/sanger-pathogens/Fastaq/pull/42) ([nds](https://github.com/nds))

## [v3.3.0](https://github.com/sanger-pathogens/Fastaq/tree/v3.3.0) (2015-03-20)
[Full Changelog](https://github.com/sanger-pathogens/Fastaq/compare/v3.2.0...v3.3.0)

**Merged pull requests:**

- Gene checks [\#41](https://github.com/sanger-pathogens/Fastaq/pull/41) ([martinghunt](https://github.com/martinghunt))

## [v3.2.0](https://github.com/sanger-pathogens/Fastaq/tree/v3.2.0) (2015-02-10)
[Full Changelog](https://github.com/sanger-pathogens/Fastaq/compare/v3.0.1...v3.2.0)

**Merged pull requests:**

- Filter handle paired files [\#40](https://github.com/sanger-pathogens/Fastaq/pull/40) ([martinghunt](https://github.com/martinghunt))
- Chunker option output one file [\#39](https://github.com/sanger-pathogens/Fastaq/pull/39) ([martinghunt](https://github.com/martinghunt))
- Chunker option output one file [\#38](https://github.com/sanger-pathogens/Fastaq/pull/38) ([martinghunt](https://github.com/martinghunt))

## [v3.0.1](https://github.com/sanger-pathogens/Fastaq/tree/v3.0.1) (2015-02-02)
[Full Changelog](https://github.com/sanger-pathogens/Fastaq/compare/v2.0.0...v3.0.1)

**Merged pull requests:**

- Rename to pyfastaq [\#37](https://github.com/sanger-pathogens/Fastaq/pull/37) ([martinghunt](https://github.com/martinghunt))
- support genetic code 4 [\#36](https://github.com/sanger-pathogens/Fastaq/pull/36) ([martinghunt](https://github.com/martinghunt))
- bug fix getting percent to keep [\#35](https://github.com/sanger-pathogens/Fastaq/pull/35) ([martinghunt](https://github.com/martinghunt))
- Add test file tasks\_test\_mean\_length.fa [\#34](https://github.com/sanger-pathogens/Fastaq/pull/34) ([martinghunt](https://github.com/martinghunt))
- Add task mean\_length [\#33](https://github.com/sanger-pathogens/Fastaq/pull/33) ([martinghunt](https://github.com/martinghunt))

## [v2.0.0](https://github.com/sanger-pathogens/Fastaq/tree/v2.0.0) (2015-01-09)
[Full Changelog](https://github.com/sanger-pathogens/Fastaq/compare/v1.7.0...v2.0.0)

**Merged pull requests:**

- Update readme for v2.0.0 [\#32](https://github.com/sanger-pathogens/Fastaq/pull/32) ([martinghunt](https://github.com/martinghunt))
- Change to one executable. Version 2.0.0 [\#31](https://github.com/sanger-pathogens/Fastaq/pull/31) ([martinghunt](https://github.com/martinghunt))
- Add script fastaq\_sort\_by\_size [\#30](https://github.com/sanger-pathogens/Fastaq/pull/30) ([martinghunt](https://github.com/martinghunt))
- Add --suffix option to fastaq\_enumerate\_names [\#29](https://github.com/sanger-pathogens/Fastaq/pull/29) ([martinghunt](https://github.com/martinghunt))

## [v1.7.0](https://github.com/sanger-pathogens/Fastaq/tree/v1.7.0) (2014-12-02)
[Full Changelog](https://github.com/sanger-pathogens/Fastaq/compare/v1.6.0...v1.7.0)

**Merged pull requests:**

- Caf to fastq [\#28](https://github.com/sanger-pathogens/Fastaq/pull/28) ([martinghunt](https://github.com/martinghunt))

## [v1.6.0](https://github.com/sanger-pathogens/Fastaq/tree/v1.6.0) (2014-11-03)
[Full Changelog](https://github.com/sanger-pathogens/Fastaq/compare/v1.5.0...v1.6.0)

**Merged pull requests:**

- v1.6.0; remove fastaq\_to\_quasr\_primers\_file; add fastaq\_sequence\_trim --revcomp [\#27](https://github.com/sanger-pathogens/Fastaq/pull/27) ([martinghunt](https://github.com/martinghunt))
- sync with martinghunt \(no changes\) [\#26](https://github.com/sanger-pathogens/Fastaq/pull/26) ([martinghunt](https://github.com/martinghunt))

## [v1.5.0](https://github.com/sanger-pathogens/Fastaq/tree/v1.5.0) (2014-10-14)
[Full Changelog](https://github.com/sanger-pathogens/Fastaq/compare/1.5.0...v1.5.0)

**Merged pull requests:**

- Script to convert python usage into manpages plus atuomatically created manpages [\#25](https://github.com/sanger-pathogens/Fastaq/pull/25) ([js21](https://github.com/js21))

## [1.5.0](https://github.com/sanger-pathogens/Fastaq/tree/1.5.0) (2014-10-10)
[Full Changelog](https://github.com/sanger-pathogens/Fastaq/compare/v1.4.1...1.5.0)

**Merged pull requests:**

- Expand nucleotides [\#24](https://github.com/sanger-pathogens/Fastaq/pull/24) ([martinghunt](https://github.com/martinghunt))

## [v1.4.1](https://github.com/sanger-pathogens/Fastaq/tree/v1.4.1) (2014-09-09)
[Full Changelog](https://github.com/sanger-pathogens/Fastaq/compare/v1.4.0...v1.4.1)

**Merged pull requests:**

- fix for fastq input [\#23](https://github.com/sanger-pathogens/Fastaq/pull/23) ([martinghunt](https://github.com/martinghunt))

## [v1.4.0](https://github.com/sanger-pathogens/Fastaq/tree/v1.4.0) (2014-09-08)
[Full Changelog](https://github.com/sanger-pathogens/Fastaq/compare/v1.3.0...v1.4.0)

**Merged pull requests:**

- Sequence trim [\#22](https://github.com/sanger-pathogens/Fastaq/pull/22) ([martinghunt](https://github.com/martinghunt))
- Added AUTHORS file [\#21](https://github.com/sanger-pathogens/Fastaq/pull/21) ([aslett1](https://github.com/aslett1))

## [v1.3.0](https://github.com/sanger-pathogens/Fastaq/tree/v1.3.0) (2014-09-03)
[Full Changelog](https://github.com/sanger-pathogens/Fastaq/compare/v1.2.0...v1.3.0)

**Merged pull requests:**

- Simulate long reads [\#20](https://github.com/sanger-pathogens/Fastaq/pull/20) ([martinghunt](https://github.com/martinghunt))

## [v1.2.0](https://github.com/sanger-pathogens/Fastaq/tree/v1.2.0) (2014-07-15)
[Full Changelog](https://github.com/sanger-pathogens/Fastaq/compare/1.1.0...v1.2.0)

## [1.1.0](https://github.com/sanger-pathogens/Fastaq/tree/1.1.0) (2014-06-10)
[Full Changelog](https://github.com/sanger-pathogens/Fastaq/compare/v1.1.0...1.1.0)

## [v1.1.0](https://github.com/sanger-pathogens/Fastaq/tree/v1.1.0) (2014-06-10)
[Full Changelog](https://github.com/sanger-pathogens/Fastaq/compare/v1.0.0...v1.1.0)

**Merged pull requests:**

- Added task to\_fasta\_union; version bump [\#19](https://github.com/sanger-pathogens/Fastaq/pull/19) ([martinghunt](https://github.com/martinghunt))

## [v1.0.0](https://github.com/sanger-pathogens/Fastaq/tree/v1.0.0) (2014-04-09)
**Merged pull requests:**

- Version 1.0.0 [\#18](https://github.com/sanger-pathogens/Fastaq/pull/18) ([martinghunt](https://github.com/martinghunt))
- Add script to find orfs [\#17](https://github.com/sanger-pathogens/Fastaq/pull/17) ([martinghunt](https://github.com/martinghunt))
- Added methods for orf finding in sequences [\#16](https://github.com/sanger-pathogens/Fastaq/pull/16) ([martinghunt](https://github.com/martinghunt))
- added option to invert filtering of sequences [\#15](https://github.com/sanger-pathogens/Fastaq/pull/15) ([martinghunt](https://github.com/martinghunt))
- Bux fix to stop using fragments shorter than the read length [\#14](https://github.com/sanger-pathogens/Fastaq/pull/14) ([martinghunt](https://github.com/martinghunt))
- added --fragments option to dump fasta file of fragments [\#13](https://github.com/sanger-pathogens/Fastaq/pull/13) ([martinghunt](https://github.com/martinghunt))
- Added option to output numbered contig names instead of using coords [\#12](https://github.com/sanger-pathogens/Fastaq/pull/12) ([martinghunt](https://github.com/martinghunt))
- Added option to output numbered contig names instead of using coords [\#11](https://github.com/sanger-pathogens/Fastaq/pull/11) ([martinghunt](https://github.com/martinghunt))
- Added option to filter file from a file of sequence names [\#10](https://github.com/sanger-pathogens/Fastaq/pull/10) ([martinghunt](https://github.com/martinghunt))
- added script fastaq\_scaffolds\_to\_contigs [\#9](https://github.com/sanger-pathogens/Fastaq/pull/9) ([martinghunt](https://github.com/martinghunt))
- Conversions to fasta - fix gff and add support for Phylip files made by Seaview [\#8](https://github.com/sanger-pathogens/Fastaq/pull/8) ([martinghunt](https://github.com/martinghunt))
- Added support for reading gbk files [\#7](https://github.com/sanger-pathogens/Fastaq/pull/7) ([martinghunt](https://github.com/martinghunt))
- Added option to omit sequences that are all Ns from fastaq\_chunker [\#6](https://github.com/sanger-pathogens/Fastaq/pull/6) ([martinghunt](https://github.com/martinghunt))
- Conversions to fasta [\#5](https://github.com/sanger-pathogens/Fastaq/pull/5) ([martinghunt](https://github.com/martinghunt))
- New script fastaq\_make\_random\_contigs; added funciotanlity to replace sections of bases within sequences [\#4](https://github.com/sanger-pathogens/Fastaq/pull/4) ([martinghunt](https://github.com/martinghunt))
- Removed debugging print [\#3](https://github.com/sanger-pathogens/Fastaq/pull/3) ([martinghunt](https://github.com/martinghunt))
- modified init file so that 'import fastaq' works; removed unecessary debug print statement [\#2](https://github.com/sanger-pathogens/Fastaq/pull/2) ([martinghunt](https://github.com/martinghunt))
- Initial add of all files [\#1](https://github.com/sanger-pathogens/Fastaq/pull/1) ([martinghunt](https://github.com/martinghunt))

